core/services/log/registrations.go

```diff
@@ -61,47 +63,50 @@ func newRegistrations() *registrations {
 	return &registrations{
 		subscribers: make(map[uint64]*subscribers),
 		decoders:    make(map[common.Address]ParseLogFunc),
+
+		// a high value for lowestNumConfirmations is set until first subscriber,
+		// to avoid a special case in comparison to its NumConfirmations
+		lowestNumConfirmations: uint64(math.MaxUint64),
 	}
 }
 
 func (r *registrations) addSubscriber(reg registration) (needsResubscribe bool) {
 	addr := reg.opts.Contract
 	r.decoders[addr] = reg.opts.ParseLog
 
-	if reg.opts.NumConfirmations == 0 {
-		reg.opts.NumConfirmations = 1
-	}
-
 	if _, exists := r.subscribers[reg.opts.NumConfirmations]; !exists {
 		r.subscribers[reg.opts.NumConfirmations] = newSubscribers()
 	}
 
 	needsResubscribe = r.subscribers[reg.opts.NumConfirmations].addSubscriber(reg)
 
-	r.maybeIncreaseHighestNumConfirmations(reg.opts.NumConfirmations)
+	r.maybeUpdateNumConfirmationsRange(reg.opts.NumConfirmations)
 	return
 }
 
 func (r *registrations) removeSubscriber(reg registration) (needsResubscribe bool) {
-	l, exists := r.subscribers[reg.opts.NumConfirmations]
+	subscribers, exists := r.subscribers[reg.opts.NumConfirmations]
 	if !exists {
 		return
 	}
 
-	needsResubscribe = l.removeSubscriber(reg)
+	needsResubscribe = subscribers.removeSubscriber(reg)
 
-	// we only need to re-evaluate highestNumConfirmations if the removed value was highest
-	if reg.opts.NumConfirmations == r.highestNumConfirmations {
-		r.resetHighestNumConfirmations()
+	if len(r.subscribers[reg.opts.NumConfirmations].handlers) == 0 {
+		delete(r.subscribers, reg.opts.NumConfirmations)
+		r.resetNumConfirmationsRange()
 	}
 	return
 }
 
-// increase the highestNumConfirmations stored if the new listener has a higher value
-func (r *registrations) maybeIncreaseHighestNumConfirmations(newNumConfirmations uint64) {
+// maybe update the numbers tracking highest and lowest num confirmations
```

Comment:
we can do better than this as a comment

Reply:
updated, and moved out of the separate method